test/functional/p2p_sendrecon.py

```diff
@@ -0,0 +1,168 @@
+        self.sendrecon_msg_received = None
+        self.manual_verack = manual_verack
+
+    def on_version(self, message):
+        if not self.manual_verack:
+            super().on_version(message)
+        self.version_received = True
+
+    def on_sendrecon(self, message):
+        print(message)
+        self.sendrecon_msg_received = message
+
+def create_sendrecon_msg():
+    sendrecon_msg = msg_sendrecon()
+    sendrecon_msg.initiator = True
+    sendrecon_msg.responder = False
+    sendrecon_msg.version = 1
+    sendrecon_msg.salt = 2
+    return sendrecon_msg
+
+class SendReconTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-txrecon']]
+
+    def run_test(self):
+        # Checks for the node *sending* SENDRECON
+        self.log.info('SENDRECON sent')
+        # Inbound
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(), send_version=True, wait_for_verack=False)
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert not peer.sendrecon_msg_received.initiator
+        assert peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+        # Outbound
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=1, connection_type="outbound-full-relay")
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert peer.sendrecon_msg_received.initiator
+        assert not peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+
+        self.log.info('SENDRECON should not be sent if block-relay-only')
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=2, connection_type="block-relay-only")
+        time.sleep(1)
+        assert not peer.sendrecon_msg_received
+
+        self.log.info('SENDRECON on pre-WTXID version should not be sent')
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(), send_version=False, wait_for_verack=False)
+        pre_wtxid_version_msg = msg_version()
+        pre_wtxid_version_msg.nVersion = 70015
+        pre_wtxid_version_msg.strSubVer = P2P_SUBVERSION
+        pre_wtxid_version_msg.nServices = P2P_SERVICES
+        pre_wtxid_version_msg.relay = 1
+        peer.send_message(pre_wtxid_version_msg)
+        time.sleep(1)
+        assert not peer.sendrecon_msg_received
+
+        # Checks for the node *receiving* SENDRECON
+        self.log.info('valid SENDRECON')
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(manual_verack=True), send_version=True, wait_for_verack=False)
+        peer.wait_until(lambda: peer.version_received)
+        peer.send_message(create_sendrecon_msg())
+        time.sleep(1)
+        assert peer.is_connected
+        self.log.info('second SENDRECON triggers a disconnect')
+        peer.send_message(create_sendrecon_msg())
+        time.sleep(1)
+        assert not peer.is_connected
```

Comment:
you can use `wait_for_disconnect()`. sleeping is insufficient; it could take longer to disconnect especially if other functional tests are being run in parallel

(same with the rest of the disconnection tests)